### PR TITLE
Read Stata version 118 files closes #9882

### LINKED ISSUE #9882: Support Stata 14 format (format 118)
Hello, I'm just leaving this here, might try to pick this up as a weekend project soon.

The beautiful people at stata have released documentation for the format here: http://www.stata.com/help.cgi?dta

And the old format is here: 
http://www.stata.com/help.cgi?dta_117

The docs say that it'll highlight changes between 118 and 117 but I couldn't find any such references. One thing I noticed is that they now have utf-8 support! Yay!

As a workaround for now, stata users who want to use stata 14 can use "saveold" which creates stata 13 files, which pandas supports.

### PR DESCRIPTION
**Pull request by kshedden**
closes #9882 

This extends the Stata reader to handle version 118 (stata 14) format files.

It correctly handles the test file posted by @makmanalp.  I don't have stata14 now to generate additional test files, but I'm not sure that it's necessary.

There is one point I'm not sure about, relating to the way that strls are indexed by a weird composite of a short and a 6 byte integer.  The stata docs aren't clear on how this 8 byte composite should (or shouldn't) be byteswapped when the file and local endianness don't match.  We currently byteswap it in this case (inheriting from our dta117 code) but I'm not sure that this is correct.

I can confirm that it works when the file and local machine are both little endian, which covers the most common situation.

I haven't looked at the writer much, it seems to always write version 114 files.

### CODE REVIEW

**Comment by jreback** [comment]
I think you should create a class for each version 
(which have a hierarchy) - introspect the file and dispatch to the class for processing 

should be simpler to maintain

**Comment by kshedden** [comment]
Having separate classes would work with ``read_stata`` but people also need to be able to create the class directly through __init__.  If someone inits the class with StataReader('myfile.dta') I don't see how we can control which subclass gets created.

**Comment by jreback** [comment]
easy, ``StataReader`` is just a holder of the work class, which is instantiated at run-time. e.g. something like

```
class StataReader(...):

    def __init__(self, file....):

           if file_is_117:
                 self.data = V117(....)
           elif file_is_118:
                 self.data = V118(....)
           else:
                 raise ValueError("cannot parse this file")

```

you would simply change the impl and call methods on the worker class, while the external API would remain the same. The big advantage would be that this would leave all of the existing classes parsed the same; adding something new should be straightforward and minimal changes.

**Comment by kshedden** [comment]
I still don't get it.  StataReader is either used as an iterator, or the user calls the read or data methods to obtain some or all of the data as a DataFrame.  StataReader also has some additional methods and attributes like variable_labels whose implementation depends on the file version.  If there were subclasses for different dta versions it seems to me that we would need to monkey patch the subclass methods into the base class (since the caller will obtain an instance of the base class not the correct subclass that matches the data file format).  Also complicating matters is the fact that we have to read partway through the file before we know what the version is.

**Comment by jreback** [comment]
just trying to make it easier / more extensibile. It seems that every time this stata format changes this needs a big rewrite.

**Comment by kshedden** [comment]
I agree that it is sort of a mess.  But the pieces that are version-specific are intertwined with code that works for all versions.  The format changes are substantial as well, so to some degree the complexity is unavoidable.

**Comment by bashtage** [comment]
> There is one point I'm not sure about, relating to the way that strls are indexed by a weird composite of a short and a 6 byte integer. The stata docs aren't clear on how this 8 byte composite should (or shouldn't) be byteswapped when the file and local endianness don't match. We currently byteswap it in this case (inheriting from our dta117 code) but I'm not sure that this is correct.

I don't think Stata runs on any big endian platforms, so the only tests that matter here are 

1. Round trip on either
2. Generate on BE, read on LE
2. Generate on LE, read on BE

**Comment by makmanalp** [comment]
Nice work! Unfortunately, I don't have access to a big-endian computer :(

File: pandas/io/tests/test_stata.py [hunk 1]
```diff
@@ -244,6 +248,24 @@ def test_read_dta12(self):
 
         tm.assert_frame_equal(parsed_117, expected, check_dtype=False)
 
+
+    def test_read_dta18(self):
+        parsed_118 = self.read_dta(self.dta22_118)
+        parsed_118["Bytes"] = parsed_118["Bytes"].astype('O')
+
+        expected = DataFrame.from_records(
+            [['Cat', 'Bogota', 'Bogotá', 1, 1.0, 'option b Ã\x9cnicode', 1.0],
```
Comment (makmanalp) [inline pandas/io/tests/test_stata.py]: Hey, is `option b Ã\x9cnicode` as intended?

Comment (kshedden) [inline pandas/io/tests/test_stata.py]: You are right  I fixed this now.

**Comment by bashtage** [comment]
Is there a minimum Stata version that is supported?  Some of the format info seems to be for very old, untested versions.  This could be cleaned up.

File: pandas/io/stata.py [hunk 2]
```diff
@@ -951,222 +953,324 @@ def __init__(self, path_or_buf, convert_dates=True,
...
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 6
+        self.seek_strls = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 7
+        self.seek_value_labels = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 14
+        self.path_or_buf.seek(seek_vartypes)
+        typlist = [struct.unpack(self.byteorder + 'H',
+                                 self.path_or_buf.read(2))[0]
+                   for i in range(self.nvar)]
+        self.typlist = [None]*self.nvar
```
Comment (jreback) [inline pandas/io/stata.py]: more idiomatic to have something like this:

```
def f(typ):
      if typ <= 2045:
            return str(typ)
      try:
            return self.DTYPE_MAP_XML[typ]
       except KeyError:
            raise ValueError("cannot convert stata dtype [{0}]"
                            .format(typ))
    
self.dtyplist = map(f, typlist)
```

File: pandas/io/stata.py [hunk 3]
```diff
@@ -951,222 +953,324 @@ def __init__(self, path_or_buf, convert_dates=True,
...
+                if typ <= 2045:
+                    self.typlist[i] = typ
+                else:
+                    self.typlist[i] = self.TYPE_MAP_XML[typ]
+                i += 1
+        except:
+            raise ValueError("cannot convert stata types [{0}]"
+                             .format(','.join(typlist)))
+        self.dtyplist = [None]*self.nvar
+        try:
```
Comment (jreback) [inline pandas/io/stata.py]: this one as well

File: pandas/io/stata.py [hunk 4]
```diff
@@ -951,222 +953,324 @@ def __init__(self, path_or_buf, convert_dates=True,
...
+        #self.path_or_buf.read(8)  # </stata_dta>
+        #self.path_or_buf.read(8)  # EOF
+        self.path_or_buf.seek(seek_vartypes)
+        typlist = [struct.unpack(self.byteorder + 'H',
+                                 self.path_or_buf.read(2))[0]
+                   for i in range(self.nvar)]
+        self.typlist = [None]*self.nvar
+        try:
+            i = 0
+            for typ in typlist:
```
Comment (jreback) [inline pandas/io/stata.py]: further you its better to make the typlist conversion as a function itself

e.g. ``self.dtypelist = self._get_dtypes_from_typelist(dtyplist)`` (then can use it in 117 and 118) versions

File: pandas/io/stata.py [hunk 5]
```diff
@@ -951,222 +953,324 @@ def __init__(self, path_or_buf, convert_dates=True,
...
+                            for i in range(self.nvar)]
+        else:
+            self.varlist = [self._null_terminate(self.path_or_buf.read(9))
+                            for i in range(self.nvar)]
+        self.srtlist = struct.unpack(
+            self.byteorder + ('h' * (self.nvar + 1)),
+            self.path_or_buf.read(2 * (self.nvar + 1))
+        )[:-1]
+        if self.format_version > 113:
+            self.fmtlist = [self._null_terminate(self.path_or_buf.read(49))
```
Comment (jreback) [inline pandas/io/stata.py]: not sure if reading of some of the older formats can share some of these utilty functions

e.g. might be useful to make functions like these

```
def _get_format_list(self):
        if self.format_version > 113:
            return [self._null_terminate(self.path_or_buf.read(49))
                            for i in range(self.nvar)]
        elif self.format_version > 104:
            return [self._null_terminate(self.path_or_buf.read(12))
                            for i in range(self.nvar)]
        else:
            return [self._null_terminate(self.path_or_buf.read(7))
                            for i in range(self.nvar)]
```

**Comment by jreback** [comment]
@kshedden I put a couple of examples of some code sharing that I think can be done (rather than inheriting from base classes, you can do it this way as well). To reduce the future maitenance burden.

File: pandas/io/tests/test_stata.py [hunk 6]
```diff
@@ -244,6 +248,38 @@ def test_read_dta12(self):
...
+        rdr = StataReader(self.dta22_118)
+        vl = rdr.variable_labels()
+        vl_expected = {u'Unicode_Cities_Strl': u'Here are some strls with Ünicode chars',
+                       u'Longs': u'long data',
+                       u'Things': u'Here are some things',
+                       u'Bytes': u'byte data',
+                       u'Ints': u'int data',
+                       u'Cities': u'Here are some cities',
+                       u'Floats': u'float data'}
+        for k in vl.keys():
```
Comment (jreback) [inline pandas/io/tests/test_stata.py]: use ``tm.assert_dict_equal``

File: pandas/io/tests/test_stata.py [hunk 7]
```diff
@@ -244,6 +248,38 @@ def test_read_dta12(self):
...
+            assert(k in vl_expected.keys())
+            assert(vl[k] == vl_expected[k])
+
+        assert(rdr.data_label == u'This is a  Ünicode data label')
```
Comment (jreback) [inline pandas/io/tests/test_stata.py]: use ``assertEqual``

File: pandas/io/stata.py [hunk 8]
```diff
@@ -1222,28 +1310,45 @@ def _read_value_labels(self):
...
+                else:
+                    buf = buf[0:2] + buf[6:]
+                v_o = struct.unpack('Q', buf)[0]
             typ = struct.unpack('B', self.path_or_buf.read(1))[0]
             length = struct.unpack(self.byteorder + 'I',
                                    self.path_or_buf.read(4))[0]
             va = self.path_or_buf.read(length)
             if typ == 130:
-                va = va[0:-1].decode(self._encoding or self._default_encoding)
+                if self.format_version == 117:
```
Comment (jreback) [inline pandas/io/stata.py]: maybe

```
encoding = 'utf-8'
if self.format_version == 117:
    encoding = self._encoding or self._default_encoding
va = va[0:-1].decode(encoding)
```

**Comment by jreback** [comment]
couple of minor comments. looks good though. after fix, pls squash. ping when green.

**Comment by kshedden** [comment]
@jreback -- this should be all set to go now.

**Comment by jreback** [comment]
@kshedden thanks! great effort on this!

**Comment by jorisvandenbossche** [comment]
I think this can still use an update in the documentation? (where they mention which versions read_stata supports) And maybe also a whatsnew note?

@kshedden can you do that in another PR?

**Comment by jorisvandenbossche** [comment]
Ah, I see @jreback  already did the whatsnew :-) 
But the docs can still use an update.

**Comment by jreback** [comment]
yes, @kshedden made add a note (or section) on compatiblity (e.g. what versions are supported) in io.rst

**Comment by makmanalp** [comment]
:cake: Thanks to all involved - this makes my life way easier!

**Pull request merged**